src/agent/api.rs

```diff
@@ -0,0 +1,139 @@
+use base64;
+use crates::{Crate, GitHubRepo};
+use errors::*;
+use ex::Experiment;
+use reqwest::{header, Client, Method, RequestBuilder, StatusCode};
+use results::TestResult;
+use serde::de::DeserializeOwned;
+use server::api_types::{AgentConfig, ApiResponse};
+use toolchain::Toolchain;
+
+trait ResponseExt {
+    fn to_api_response<T: DeserializeOwned>(self) -> Result<T>;
+}
+
+impl ResponseExt for ::reqwest::Response {
+    fn to_api_response<T: DeserializeOwned>(mut self) -> Result<T> {
+        // 404 responses are not JSON, so avoid parsing them
+        if self.status() == StatusCode::NotFound {
+            bail!("invalid API enpoint called");
+        }
+
+        let result: ApiResponse<T> = self.json().chain_err(|| "failed to parse API response")?;
+        match result {
+            ApiResponse::Success { result } => Ok(result),
+            ApiResponse::InternalError { error } => bail!("internal server error: {}", error),
+            ApiResponse::Unauthorized => bail!("invalid authorization token provided"),
+        }
+    }
+}
+
+const RETRY_AFTER: u64 = 5;
+
+pub struct AgentApi {
+    client: Client,
+    url: String,
+    token: String,
+}
+
+impl AgentApi {
+    pub fn new(url: &str, token: &str) -> Self {
+        AgentApi {
+            client: Client::new(),
+            url: url.to_string(),
+            token: token.to_string(),
+        }
+    }
+
+    fn build_request(&self, method: Method, url: &str) -> RequestBuilder {
+        let mut req = self.client
+            .request(method, &format!("{}/agent-api/{}", self.url, url));
+        req.header(header::Authorization(format!("token {}", self.token)));
```

Comment:
You've defined `impl Scheme for Token` in `src/server/auth.rs` - why not use `Token` here?

Reply:
Forgot to do that after refactoring to use a struct. Fixed.